myia/cconv.py

```diff
@@ -0,0 +1,165 @@
+"""
+Utilities for closure conversion.
+
+I.e. retrieving nesting structure and listing free variables.
+"""
+
+
+from typing import Set, Dict, Optional, Union
+from collections import defaultdict
+from .anf_ir import ANFNode, Constant, Graph
+from .anf_ir_utils import dfs
+
+
+def is_graph(x: ANFNode) -> bool:
+    """Return whether x is a Constant with a Graph value."""
+    return isinstance(x, Constant) and isinstance(x.value, Graph)
+
+
+class ParentProxy:
+    """Represents a graph's immediate parent."""
+
+    def __init__(self, graph: Graph) -> None:
+        """Initialize the ParentProxy."""
+        self.graph = graph
+
+
+class NestingAnalyzer:
+    """
+    Analyzes the nesting structure of a graph.
+
+    Call the `run(root_graph)` method to populate the data structures.
+
+    Attributes:
+        deps: Maps a graph to the set of graphs that it depends on.
+        parents: Maps a graph to its immediate parent or `None` if
+            it has no parent.
+        fvs: Maps a graph to a set of nodes it refers to from other
+            graphs, including nodes nested graphs have to access.
+
+    """
+
+    def __init__(self) -> None:
+        """Initialize the NestingAnalyzer."""
+        self.processed: Set[Graph] = set()
+        self.parents: Dict[Graph, Optional[Graph]] = {}
+        self.fvs: Dict[Graph, Set[ANFNode]] = defaultdict(set)
+
+    def run(self, root: Graph) -> 'NestingAnalyzer':
+        """Run the analysis. Populates `self.parents` and `self.fvs`."""
+        if root in self.processed:
+            return self
+
+        # Initialize temporary structures
+        self.prox: Dict[Graph, ParentProxy] = {}
+        self.graphs: Set[Graph] = set()
+        self.nodes: Set[ANFNode] = set()
+        self.deps: Dict[Graph, Set[Union[Graph, ParentProxy]]] = \
+            defaultdict(set)
+
+        # We get all nodes that could possibly be executed within this
+        # graph.
+        self.nodes = set(dfs(Constant(root), True))
+
+        # We get all graphs possibly accessed.
+        self.graphs = {node.value for node in self.nodes if is_graph(node)}
+        self.processed |= self.graphs
+
+        # We compute the dependencies:
+        # * If a node in graph g has an input from graph g', then g depends
+        #   on g'
+        # * If a node in graph g refers to a graph g', then g depends on
```

Comment:
Is this true? For example:
```python
def f(x):
    return x

def g():
    return f
```
I think you only need to look at the free variables.

Reply:
It's true: the minimal possible nesting of `g` is the minimal nesting of `f`'s parent, which in this case happens to be `global` (if it was nested below, it wouldn't be able to see `g`). I suppose the main reason I did it this way is that it allows part of the algorithm to be run non-recursively and it made it easier for me to see what's going on (some free variables belong to no graphs, because they *are* graphs that we have yet to place in the hierarchy, and that confused me a bit at first). All this being said, I'm pretty sure it can be simplified, if need be.